zerver/webhooks/gocd/view.py

```diff
@@ -26,31 +39,75 @@ def api_gocd_webhook(
-
-    if result == "Passed":
-        emoji = ":thumbs_up:"
-    elif result == "Failed":
-        emoji = ":thumbs_down:"
-
-    build_details_file = os.path.join(os.path.dirname(__file__), "fixtures/build_details.json")
-
-    with open(build_details_file) as f:
-        contents = json.load(f)
-        build_link = contents["build_details"]["_links"]["pipeline"]["href"]
-
-    body = MESSAGE_TEMPLATE.format(
-        modifications["user_name"].tame(check_string),
-        result,
-        emoji,
-        build_link,
-        modifications["comment"].tame(check_string),
+    type = payload["type"].tame(check_string)
+    if type == "stage":
+        body = get_body(payload)
+        topic_name = get_topic(payload)
+        check_send_webhook_message(request, user_profile, topic_name, body)
+    return json_success(request)
+
+
+def get_topic(payload: WildValue) -> str:
+    return TOPIC_TEMPLATE.format(
+        pipeline=payload["data"]["pipeline"]["name"].tame(check_string),
+        stage=payload["data"]["pipeline"]["stage"]["name"].tame(check_string),
     )
-    branch = material["description"].tame(check_string).split(",")
-    topic_name = branch[0].split(" ")[1]
 
-    check_send_webhook_message(request, user_profile, topic_name, body)
 
-    return json_success(request)
+def get_commit_details(payload: WildValue) -> str:
+    material = payload["data"]["pipeline"]["build-cause"][0]["material"]
+    url = material["git-configuration"]["url"].tame(check_string)
+    revision = payload["data"]["pipeline"]["build-cause"][0]["modifications"][0]["revision"].tame(
+        check_string
+    )
+    commit_sha = get_short_sha(revision)
+    url = f"{url}/commit/{commit_sha}"
+    branch = material["git-configuration"]["branch"].tame(check_string)
+    return COMMIT_INFO_TEMPLATE.format(
+        commit_details=commit_sha,
+        commit_link=url,
+        branch_name=branch,
+    )
+
+
+def get_jobs_details(pipeline_data: WildValue) -> str:
+    job_dict_list = pipeline_data["stage"]["jobs"]
+    formatted_job_dict = defaultdict(list)
+    job_details_template = ""
+
+    for job in job_dict_list:
+        job_name = job["name"].tame(check_string)
+        job_result = job["result"].tame(check_string)
+        formatted_job_dict[job_result].append(f"`{job_name}`")
+
+    for key in formatted_job_dict:
+        formatted_job_list = ", ".join(formatted_job_dict[key])
+        job_details_template += f"- **{key} job(s)**: {formatted_job_list}\n"
+
+    return job_details_template
+
+
+def get_body(payload: WildValue) -> str:
+    pipeline_data = payload["data"]["pipeline"]
+    body_details = {
+        "commit_details": get_commit_details(payload),
+        "status": pipeline_data["stage"].get("state").tame(check_string),
+        "pipeline": pipeline_data["name"].tame(check_string),
+        "stage": pipeline_data["stage"]["name"].tame(check_string),
+        "start_time": pipeline_data["stage"]["create-time"].tame(check_string),
```

Comment:
This would be easier for me to read if we were consistent between using `.get()` or just the  `["key-1"]["key-2"]` format here and below.

Reply:
Done
```diff
@@ -91,7 +88,7 @@ def get_body(payload: WildValue) -> str:
     pipeline_data = payload["data"]["pipeline"]
     body_details = {
         "commit_details": get_commit_details(payload),
-        "status": pipeline_data["stage"].get("state").tame(check_string),
+        "status": pipeline_data["stage"]["state"].tame(check_string),
         "pipeline": pipeline_data["name"].tame(check_string),
         "stage": pipeline_data["stage"]["name"].tame(check_string),
         "start_time": pipeline_data["stage"]["create-time"].tame(check_string),
@@ -99,12 +96,12 @@ def get_body(payload: WildValue) -> str:
     if body_details["status"] == "Building":
         return SCHEDULED_BODY_TEMPLATE.format(**body_details)
     else:
-        result = pipeline_data["stage"].get("result").tame(check_string)
+        result = pipeline_data["stage"]["result"].tame(check_string)
         body_details.update(
             {
                 "result": result,
                 "emoji": ":check:" if result == "Passed" else ":warning:",
-                "end_time": pipeline_data["stage"].get("last-transition-time").tame(check_string),
+                "end_time": pipeline_data["stage"]["last-transition-time"].tame(check_string),
             }
         )

```

Reply:
So, those keys might not be present? Then I think we need the `check_none_or` in that case, right?

Reply:
oops, I meant it's probably ok to use ["key1"]["key2"] as those keys exist in all known payload types